tests/contrib/flask/test_signals.py

```diff
@@ -1,57 +1,176 @@
-
-from flask import Flask
-
-
-class FlaskBlinkerCase(TestCase):
-    """Ensures that the integration between Flask and Blinker
-    to trace Flask endpoints works as expected
-    """
-    def get_app(self):
-        """Creates a new Flask App"""
-        app = Flask(__name__)
-
-        # add testing routes here
-        @app.route('/')
-        def index():
-            return 'Hello world!'
-
-        return app
-
-    def setUp(self):
-        # initialize a traced app with a dummy tracer
-        app = self.get_app()
-        self.tracer = get_dummy_tracer()
-        self.traced_app = TraceMiddleware(app, self.tracer)
-
-        # make the app testable
-        app.config['TESTING'] = True
-        self.app = app.test_client()
-
-    def test_signals_without_weak_references(self):
-        # it should work when the traced_app reference is not
-        # stored by the user and the garbage collection starts
-        self.traced_app = None
-        gc.collect()
-
-        r = self.app.get('/')
-        eq_(r.status_code, 200)
-
-        traces = self.tracer.writer.pop_traces()
-        eq_(len(traces), 1)
-        eq_(len(traces[0]), 1)
-
-        span = traces[0][0]
-        eq_(span.service, 'flask')
-        eq_(span.name, 'flask.request')
-        eq_(span.span_type, 'http')
-        eq_(span.resource, 'index')
-        eq_(span.get_tag('http.status_code'), '200')
-        eq_(span.get_tag('http.url'), 'http://localhost/')
-        eq_(span.error, 0)
+import mock
+
+import flask
+
+from ddtrace import Pin
+from ddtrace.contrib.flask import unpatch
+from ddtrace.contrib.flask.patch import flask_version
+
+from . import BaseFlaskTestCase
+
+
+class FlaskSignalsTestCase(BaseFlaskTestCase):
+    def get_signal(self, signal_name):
+        # v0.9 missed importing `appcontext_tearing_down` in `flask/__init__.py`
+        #  https://github.com/pallets/flask/blob/0.9/flask/__init__.py#L35-L37
+        #  https://github.com/pallets/flask/blob/0.9/flask/signals.py#L52
+        # DEV: Version 0.9 doesn't have a patch version
+        if flask_version <= (0, 9) and signal_name == 'appcontext_tearing_down':
+            return getattr(flask.signals, signal_name)
+        return getattr(flask, signal_name)
+
+    def signal_function(self, name):
+        def signal(*args, **kwargs):
+            pass
+
+        func = mock.Mock(signal, name=name)
+        func.__module__ = 'tests.contrib.flask'
+        func.__name__ = name
+        return func
+
+    def call_signal(self, signal_name, *args, **kwargs):
+        """Context manager helper used for generating a mock signal function and registering with flask"""
+        func = self.signal_function(signal_name)
+
+        signal = self.get_signal(signal_name)
+        signal.connect(func, self.app)
+
+        try:
+            signal.send(*args, **kwargs)
+            return func
+        finally:
+            # DEV: There is a bug in `blinker.Signal.disconnect` :(
```

Comment:
Is there a link to a bug report that we can reference here?

Reply:
there isn't, I should probably open one.

the TL;DR; it has something to do with how they create weak references and when you disconnect they can't find the original function in their list... something like that.